File: server/schemas/layout.py
from typing import Dict, List
from pydantic import BaseModel

class Position(BaseModel):
    x: float
    y: float

class RectangleRange(BaseModel):
    lu: Position
    rd: Position

class Density(BaseModel):
    xs: List[int]
    densities: List[int]
 
Layout = Dict[str, Position]

# class LayoutNode(BaseModel):
#     segment: str
#     position: Position
#     # TODO: bounds
#     # xl: float
#     # xr: float

File: src/types/layout.ts
export interface Position {
  x: number;
  y: number;
}

export interface Bounds {
  xl: number;
  xr: number;
}

// export interface LayoutNode {
//   segment: string;
//   position: Position;
// }

export interface RectangleRange {
  lu: Position;
  rd: Position;
}

export type Layout = Record<string, Position>;

export interface GraphLayoutState {
  viewport: RectangleRange;
  zoom: number;
  pan: Position;
}